File: android-headunit/app/src/main/java/com/mobilelabkit/headunit/AapProto.kt
package com.mobilelabkit.headunit

/**
 * Wire constants for the (modern) Android Auto protocol.
 *
 * Framing on the bulk link is unchanged from before:
 *   [channelId:1][flags:1][frameSize:2 BE] (+[totalSize:4 BE] when flags has FIRST) [payload]
 * flags = frameType | encryptionType | messageType. The decrypted payload starts with a
 * 2-byte BE message id, then the protobuf.
 *
 * Message ids + message shapes come from the vendored modern protos
 * (com.andrerinas.headunitrevived.aap.protocol.proto.*) — see Control/Media/Sensors/Input.
 * aasdk's 2018 message set is too old for Android Auto 1.7.
 */
object AapProto {
    // --- channel ids (headunit-revived Channel.kt; head unit assigns these) ---
    const val CH_CONTROL = 0
    const val CH_SENSOR = 1
    const val CH_VIDEO = 2
    const val CH_INPUT = 3
    const val CH_AUDIO_SPEECH = 4   // AU1
    const val CH_AUDIO_SYSTEM = 5   // AU2
    const val CH_AUDIO_MEDIA = 6    // AUD
    const val CH_MIC = 7
    const val CH_BLUETOOTH = 8
    const val CH_MEDIA_PLAYBACK = 9
    const val CH_NAV = 10

    fun isAudio(ch: Int) = ch == CH_AUDIO_SPEECH || ch == CH_AUDIO_SYSTEM || ch == CH_AUDIO_MEDIA
    fun isMediaLike(ch: Int) = ch == CH_VIDEO || ch == CH_MIC || isAudio(ch)

    // --- flags byte components ---
    const val FRAME_MIDDLE = 0
    const val FRAME_FIRST = 1
    const val FRAME_LAST = 2
    const val FRAME_BULK = 3
    const val FRAME_TYPE_MASK = 3
    const val ENC_PLAIN = 0
    const val ENC_ENCRYPTED = 1 shl 3 // 0x08
    const val MSG_SPECIFIC = 0
    const val MSG_CONTROL = 1 shl 2   // 0x04

    // Protocol version we advertise (major.minor).
    const val VERSION_MAJOR = 1
    const val VERSION_MINOR = 1

    // Control message ids that aren't the SSL/version raw path (rest come from Control proto).
    const val VERSION_REQUEST = 1
    const val VERSION_RESPONSE = 2
    const val SSL_HANDSHAKE = 3       // MESSAGE_ENCAPSULATED_SSL
    const val AUTH_COMPLETE = 4

    const val MAX_FRAME_PAYLOAD = 0x4000

    fun channelName(id: Int): String = when (id) {
        CH_CONTROL -> "CONTROL"; CH_SENSOR -> "SENSOR"; CH_VIDEO -> "VIDEO"; CH_INPUT -> "INPUT"
        CH_AUDIO_SPEECH -> "AUD_SPEECH"; CH_AUDIO_SYSTEM -> "AUD_SYSTEM"; CH_AUDIO_MEDIA -> "AUD_MEDIA"
        CH_MIC -> "MIC"; CH_BLUETOOTH -> "BT"; CH_MEDIA_PLAYBACK -> "MEDIA_PB"; CH_NAV -> "NAV"
        else -> "CH$id"
    }
}

File: ios-headunit/AAHeadUnitCore/Sources/AAHeadUnitCore/AapProto.swift
import Foundation

/// Wire constants for the (modern) Android Auto protocol — the Swift twin of the working
/// Kotlin head unit's `AapProto.kt`. Framing on the link:
///
///   [channelId:1][flags:1][frameSize:2 BE] (+[totalSize:4 BE] when flags has FIRST) [payload]
///
/// flags = frameType | encryptionType | messageType. The decrypted payload starts with a
/// 2-byte BE message id, then the protobuf.
public enum AapProto {
    // Channel ids (head unit assigns these).
    public static let chControl = 0
    public static let chSensor = 1
    public static let chVideo = 2
    public static let chInput = 3
    public static let chAudioSpeech = 4
    public static let chAudioSystem = 5
    public static let chAudioMedia = 6
    public static let chMic = 7
    public static let chBluetooth = 8
    public static let chMediaPlayback = 9
    public static let chNav = 10

    public static func isAudio(_ ch: Int) -> Bool {
        ch == chAudioSpeech || ch == chAudioSystem || ch == chAudioMedia
    }
    public static func isMediaLike(_ ch: Int) -> Bool {
        ch == chVideo || ch == chMic || isAudio(ch)
    }

    // flags byte components
    public static let frameMiddle = 0
    public static let frameFirst = 1
    public static let frameLast = 2
    public static let frameBulk = 3
    public static let frameTypeMask = 3
    public static let encPlain = 0
    public static let encEncrypted = 1 << 3   // 0x08
    public static let msgSpecific = 0
    public static let msgControl = 1 << 2      // 0x04

    public static let versionMajor = 1
    public static let versionMinor = 2

    // Control message ids on the raw (pre-protobuf) path.
    public static let versionRequest = 1
    public static let versionResponse = 2
    public static let sslHandshake = 3         // MESSAGE_ENCAPSULATED_SSL
    public static let authComplete = 4

    public static let maxFramePayload = 0x4000

    public static func channelName(_ id: Int) -> String {
        switch id {
        case chControl: return "CONTROL"
        case chSensor: return "SENSOR"
        case chVideo: return "VIDEO"
        case chInput: return "INPUT"
        case chAudioSpeech: return "AUD_SPEECH"
        case chAudioSystem: return "AUD_SYSTEM"
        case chAudioMedia: return "AUD_MEDIA"
        case chMic: return "MIC"
        case chBluetooth: return "BT"
        case chMediaPlayback: return "MEDIA_PB"
        case chNav: return "NAV"
        default: return "CH\(id)"
        }
    }

    // Big-endian helpers.
    public static func u16be(_ v: Int) -> [UInt8] { [UInt8((v >> 8) & 0xff), UInt8(v & 0xff)] }
    public static func u32be(_ v: Int) -> [UInt8] {
        [UInt8((v >> 24) & 0xff), UInt8((v >> 16) & 0xff), UInt8((v >> 8) & 0xff), UInt8(v & 0xff)]
    }
    public static func readU16(_ b: [UInt8], _ o: Int) -> Int {
        (Int(b[o]) << 8) | Int(b[o + 1])
    }
    public static func readU32(_ b: [UInt8], _ o: Int) -> Int {
        (Int(b[o]) << 24) | (Int(b[o + 1]) << 16) | (Int(b[o + 2]) << 8) | Int(b[o + 3])
    }
}
